omnigibson/robots/behaviorbot.py

```diff
@@ -0,0 +1,677 @@
+            if ctrl_type == ControlType.EFFORT:
+                joint.set_effort(ctrl, normalized=norm)
+            elif ctrl_type == ControlType.VELOCITY:
+                joint.set_vel(ctrl, normalized=norm, drive=True)
+            elif ctrl_type == ControlType.POSITION:            
+                if "rx" in joint.joint_name or "ry" in joint.joint_name or "rz" in joint.joint_name: 
+                    joint.set_pos(ctrl, normalized=norm, drive=False)
+                else:
+                    joint.set_pos(ctrl, normalized=norm, drive=True)
+            elif ctrl_type == ControlType.NONE:
+                # Do nothing
+                pass
+            else:
+                raise ValueError("Invalid control type specified: {}".format(ctrl_type))
+
+            # Finally, increment the current index based on how many DOFs were just controlled
+            cur_indices_idx += joint_dof
+
+    def teleop_data_to_action(self, teleop_data: dict) -> np.ndarray:
+        """
+        Generates an action for the Behaviorbot to perform based on vr data dict.
+
+        Action space (all non-normalized values that will be clipped if they are too large)
+        Body:
+        - 6DOF pose - relative to body frame from previous frame
+        Eye:
+        - 6DOF pose - relative to body frame (where the body will be after applying this frame's action)
+        Left hand, right hand (in that order):
+        - 6DOF pose - relative to body frame (same as above)
+        - 10DOF gripper joint rotation 
+
+        Total size: 44
+        """
+        # Actions are stored as 1D numpy array
+        action = np.zeros(self.action_dim)
+        # Update body action space
+        hmd_pos, hmd_orn = teleop_data["transforms"]["hmd"]
+        if np.all(np.equal(hmd_pos, np.zeros(3))) and np.all(np.equal(hmd_orn, np.array([0, 0, 0, 1]))):
+            des_body_pos, des_body_orn = self.get_position_orientation()
+            des_body_rpy = R.from_quat(des_body_orn).as_euler("XYZ")
+        else:
+            des_body_pos = hmd_pos - np.array([0, 0, 0.45])
+            des_body_rpy = np.array([0, 0, R.from_quat(hmd_orn).as_euler("XYZ")[2]])
+            des_body_orn = T.euler2quat(des_body_rpy)
+        action[self.controller_action_idx["base"]] = np.r_[des_body_pos, des_body_rpy]
+        # Update action space for other VR objects
+        for part_name, eef_part in self.parts.items():
+            # Process local transform adjustments
+            prev_local_pos, prev_local_orn = eef_part.local_position_orientation
+            reset = 0
+            hand_data = None
+            if part_name == "head":
+                part_pos, part_orn = teleop_data["transforms"]["hmd"]
+            else:
+                hand_name, hand_index = ("left", 0) if part_name == "lh" else ("right", 1)
+                if "hand_data" in teleop_data:
+                    if "raw" in teleop_data["hand_data"][hand_name]:
```

Comment:
hmmm more string matching here, seems brittle.

Reply:
make it macro

Reply:
changed to dataclass to get rid of strings